### PR TITLE
feat: add Collection#ensure

### PR DESCRIPTION
**Pull request by kxmndz**
Inspired by [Enmap](https://enmap.evie.dev)`#ensure`, `Collection#ensure` gets an element with the specified key if it exists, otherwise sets it to the provided `defaultValue` and returns the `defaultValue`. Useful for things like per-guild settings where you want to get it or set it to a default value all in one call.

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -66,6 +66,18 @@ export class Collection<K, V> extends Map<K, V> {
...
 	}
 
+	/**
+	 * Gets an element if the key exists, otherwise sets it to {@param defaultValue} and returns the {@param defaultValue}.
+	 * @param {*} key - Key to get from/set to the collection.
+	 * @param {*} defaultValue - Default value to be set and returned if the key doesn't exist.
+	 * @returns {*}
+	 */
+	public ensure(key: K, defaultValue: V): V {
+		if (this.has(key)) return this.get(key)!;
```
Comment (kyranet) [inline src/index.ts:76]: You can skip a second map scan:
```suggestion
		const value = this.get(key);
		if (typeof value !== 'undefined') return value;
```

**Review by kyranet — APPROVED** [review]

**Review by kyranet — CHANGES_REQUESTED** [review]
(Wrong button, sorry) ^

File: src/index.ts [hunk 2]
```diff
@@ -66,6 +66,19 @@ export class Collection<K, V> extends Map<K, V> {
 		return super.clear();
 	}
 
+	/**
+	 * Gets an element if the key exists, otherwise sets it to {@param defaultValue} and returns the {@param defaultValue}.
```
Comment (kyranet) [inline src/index.ts:70]:
```suggestion
	 * Gets an element if the key exists, otherwise sets and returns {@param defaultValue}.
```

File: src/index.ts [hunk 3]
```diff
@@ -66,6 +66,19 @@ export class Collection<K, V> extends Map<K, V> {
...
+	 * @param {*} key - Key to get from/set to the collection.
+	 * @param {*} defaultValue - Default value to be set and returned if the key doesn't exist.
+	 * @returns {*}
```
Comment (kyranet) [inline src/index.ts:73]:
```suggestion
	 * @returns {*} The existing value if any, {@param defaultValue} otherwise.
```

File: src/index.ts [hunk 4]
```diff
@@ -66,6 +66,19 @@ export class Collection<K, V> extends Map<K, V> {
 		return super.clear();
 	}
 
+	/**
+	 * Gets an element if the key exists, otherwise sets it to {@param defaultValue} and returns the {@param defaultValue}.
+	 * @param {*} key - Key to get from/set to the collection.
```
Comment (kyranet) [inline src/index.ts:71]:
```suggestion
	 * @param {*} key - Key to ensure from the collection.
```

Although I'm not 100% about this wording, cc: @vladfrangu @iCrawl

**Review by kyranet — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 4 above)
Comment (kxmndz) [reply src/index.ts:71]: I was thinking something closer to "Key to ensure *in* the collection", because this method ensures that the key exists in the collection

File: src/index.ts [hunk 5]
```diff
@@ -66,6 +66,19 @@ export class Collection<K, V> extends Map<K, V> {
...
 
+	/**
+	 * Gets an element if the key exists, otherwise sets and returns {@param defaultValue}.
+	 * @param {*} key - Key to get from/set to the collection.
+	 * @param {*} defaultValue - Default value to be set and returned if the key doesn't exist.
+	 * @returns {*}
+	 */
+	public ensure(key: K, defaultValue: V): V {
+		const value = this.get(key);
+		if (typeof value !== 'undefined') return value;
```
Comment (1Computer1) [inline src/index.ts:77]: This is incorrect if V extends undefined.

**Review by 1Computer1 — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 6]
```diff
@@ -66,6 +66,19 @@ export class Collection<K, V> extends Map<K, V> {
 		return super.clear();
 	}
 
+	/**
+	 * Gets an element if the key exists, otherwise sets and returns {@param defaultValue}.
+	 * @param {*} key - Key to get from/set to the collection.
+	 * @param {*} defaultValue - Default value to be set and returned if the key doesn't exist.
+	 * @returns {*} The existing value if any, {@param defaultValue} otherwise.
+	 */
+	public ensure(key: K, defaultValue: V): V {
```
Comment (1Computer1) [inline src/index.ts:75]: Consider either making this `() => V` or another function with that instead for lazily evaluating the default.

File: src/index.ts (hunk 5 above)
Comment (kxmndz) [reply src/index.ts]: I did some testing, `typeof null` is an Object and `typeof NaN` is a Number, so unless they manually set undefined as a value I don't think we have to worry about that.

Comment (1Computer1) [reply src/index.ts]: Yes, that is why this is incorrect.

File: src/index.ts (hunk 6 above)
Comment (1Computer1) [reply src/index.ts]: Your change does not solve this. `defaultValue` is evaluated at the function call, not when used. It should look like `coll.ensure(k, () => v)`.

File: src/index.ts (hunk 1 above)
Comment (DTrombett) [reply src/index.ts]: Mmh what if the the collection has that key but its value is `undefined`?

File: src/index.ts (hunk 6 above)
Comment (kxmndz) [reply src/index.ts]: My bad, misread default as result.

File: src/index.ts (hunk 1 above)
Comment (kxmndz) [reply src/index.ts]: I agree, I think an additional scan to guard against this is preferable over saving a couple of milliseconds and potentially setting the default instead of recognizing that the key already exists in the collection.

File: src/index.ts [hunk 7]
```diff
@@ -66,6 +66,37 @@ export class Collection<K, V> extends Map<K, V> {
...
+	 * client.on('guildMemberAdd', member => {
+	 *   const getGuildSettings = client.settings.ensure(message.guild.id, () => defaultSettings);
+	 *
+	 *   // use the data
+	 *   if (guildSettings.welcomeChannelID) {
+	 *     const channel = client.channels.fetch(guildSettings.welcomeChannelID);
+	 *     channel.send(`Hello <@${member.id}>, welcome to ${member.guild.name}! Enjoy your stay!`);
+	 *   }
+	 * })
+	 */
```
Comment (iCrawl) [inline src/index.ts:92]: This example is too big.

It also uses a lot of discord.js stuff that is not really related to a generic collection library. Preferably we just remove this whole example. The docstring is explaining well enough what this method does.

**Review by iCrawl — CHANGES_REQUESTED** [review]
We should think about renaming this to `upsert` instead of `ensure`.

File: src/index.ts (hunk 7 above)
Comment (kxmndz) [reply src/index.ts]: Upsert seems like it should be a separate method since they do different things. From what I understand, at least, upsert either inserts a value or updates it, and ensure is to make sure that there's an entry for that key *without* changing existing entries.

**Comment by ImRodry** [comment]
Is this PR still being maintained @kxmndz?

**Comment by kxmndz** [comment]
> Is this PR still being maintained @kxmndz?

Yes, it's ready for review @ImRodry

**Comment by ImRodry** [comment]
Then it should be rebased and the reviews should be addressed

**Comment by kxmndz** [comment]
I've decided to just close this, it was just an old thing I wanted to do and I don't have the time right now to maintain it. If someone else wants to take over though feel free.
